opencti-platform/opencti-graphql/src/manager/retentionManager.ts

```diff
@@ -77,17 +80,22 @@ const executeProcessing = async (context: AuthContext, retentionRule: RetentionR
   const result = await getElementsToDelete(context, scope, before, filters);
   const remainingDeletions = result.pageInfo.globalCount;
   const elements = result.edges;
-  logApp.debug(`[OPENCTI] Retention manager clearing ${elements.length} elements`);
-  for (let index = 0; index < elements.length; index += 1) {
-    const { node } = elements[index];
-    const { updated_at: up } = node;
-    const humanDuration = moment.duration(utcDate(up).diff(utcDate())).humanize();
-    try {
-      await deleteElement(context, scope, scope === 'knowledge' ? node.internal_id : node.id, node.entity_type);
-      logApp.debug(`[OPENCTI] Retention manager deleting ${node.id} after ${humanDuration}`);
-    } catch (e) {
-      logApp.error(e, { id: node.id, manager: 'RETENTION_MANAGER' });
-    }
+  if (elements.length > 0) {
+    logApp.debug(`[OPENCTI] Retention manager clearing ${elements.length} elements`);
+    const start = new Date().getTime();
+    const deleteFn = async (element: BasicStoreCommonEdge<StoreObject>) => {
+      const { node } = element;
+      const { updated_at: up } = node;
+      try {
+        const humanDuration = moment.duration(utcDate(up).diff(utcDate())).humanize();
+        await deleteElement(context, scope, scope === 'knowledge' ? node.internal_id : node.id, node.entity_type);
+        logApp.debug(`[OPENCTI] Retention manager deleting ${node.id} after ${humanDuration}`);
+      } catch (e) {
+        logApp.error(e, { id: node.id, manager: 'RETENTION_MANAGER' });
+      }
+    };
+    await BluePromise.map(elements, deleteFn, { concurrency: RETENTION_MAX_CONCURRENCY });
```

Comment:
Just for my understanding: if we batch elements that are interdependent, the first one might take a lock on another one in the concurrent batch, and when released the second element to delete might already be deleted. AM I right ?

If so, not harmful but we'll end up with a bunch of error logs.

Reply:
Add code to prevent logging for an already deleted error